add personal bests onto classification table

LoganHeasman200 opened the pull request:
adding personal bests onto classification table, 

- Had to change the way the max score was calculated as metric rounds were showing NaN due to logic trying to multiply number of arrows with 'X' so just comparing the max possible arrow score and if it's an X changing it to a 10.
- Wrapped classification table in a details/summary as per suggestion from Ant 
- added calculatePersonalBest function to search through local storage for the highest score for that round and defaulting to 0 if no score has been submitted 
- changed how the roundScores are calculated due to using index made getting the correct label near impossible as rounds have differing numbers of classifications available, changed from a for loop using index to an array map using the id only downside is this needs to be subtracted by one as id's are not 0-indexed 

![image](https://github.com/user-attachments/assets/5fd1400b-6493-4905-9453-972f971d091f)

quii commented:
Hey, nice, but have a couple of nits im afraid

> Had to change the way the max score was calculated as metric rounds were showing NaN due to logic trying to multiply number of arrows with 'X' so just comparing the max possible arrow score and if it's an X changing it to a 10.

I think you've duplicated some of the logic of `convertToValue`. Use this instead, and it'll work for Worcester rounds too. 

https://github.com/quii/vue-fast/blob/main/src/domain/scores.js

For `calculatePersonalBest`, I like that you put it in `domain`. But I'm not too fond of the domain code depending on vue (dependency inversion etc etc), so could you instead pass in the history to the function _from_ vue. That would also make it easier for you to write a unit test or two for the function too (please write a unit test or two for the function :p)

quii commented:
Also it seems your change has broken some tests, remember to run `npm test`

LoganHeasman200 commented:
I just realised as well that I don't even need the function to calculatePersonalBest as you've already done it inside the historystore I just didn't realise until I went poking around in the vuedebugtools

quii commented:
Ah yeah of course, that's how it gets highlighted in the history list

LoganHeasman200 commented:
ok hopefully I'll be able to see something tomorrow because this failing test is one of the worst kind the "intermittent" kind, it only fails on initial load after refresh it doesn't happen and I can't see where the issue is coming from

quii commented:
Which test is intermittent for you? It works on my machine™️ (but also the github pipeline tbf) very consistently

LoganHeasman200 commented:
[7m FAIL [27m src/domain/classification.test.js[2m > calculateClassification[2m > calculates what classification you got from a score
AssertionError: expected 'PB' to deeply equal 'U/C'

It's the same as the github workflow but I think I may have an idea of why just need to check later

LoganHeasman200 commented:
added in what should fix it, i forgot that in the section that is calculating your classification because im synthetically creating a PB in the roundscores object the function is always then returning PB as the achieved classification if you haven't achieved one so i've updated the find predicates to only find rounds where the score is greater than the score given AND the rounds name is NOT 'PB' I've also updated the tests to expect to receive the PB in it's response

LoganHeasman200 commented:
i just realised as i wrote that last comment that I shouldn't be creating a synthetic roundscores entry if there's no personal best set so wrapped it in an if statement this also removes the line when you haven't submitted that round before and the PB would just say 0 achieved, this is the preferable method

quii commented:
Good job. One for later, i think it would be better if the classification table ordered by score, so then your PB could sit in between whatever classification in lands, which makes sense (to me)

LoganHeasman200 commented:
Completely right i thought of that at first but didn't do it because of the indexing thing I spoke about earlier, but since it's array map it'll be much easier to do that now will make a note to come back to that another day!
